File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/service/impl/UseServiceImpl.java
package com.windea.demo.cloudcollect.core.service.impl;

import com.windea.commons.kotlin.extension.RandomExtension;
import com.windea.demo.cloudcollect.core.domain.entity.*;
import com.windea.demo.cloudcollect.core.domain.enums.Role;
import com.windea.demo.cloudcollect.core.domain.model.JwtUserDetails;
import com.windea.demo.cloudcollect.core.domain.view.EmailRegisterView;
import com.windea.demo.cloudcollect.core.domain.view.UsernamePasswordLoginView;
import com.windea.demo.cloudcollect.core.exception.NotFoundException;
import com.windea.demo.cloudcollect.core.repository.*;
import com.windea.demo.cloudcollect.core.service.EmailService;
import com.windea.demo.cloudcollect.core.service.UserService;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;

@Service
public class UseServiceImpl implements UserService {
	private final UserRepository repository;
	private final CollectRepository collectRepository;
	private final CollectCategoryRepository categoryRepository;
	private final NoticeRepository noticeRepository;
	private final EmailService emailService;
	private final PasswordEncoder passwordEncoder;
	private final AuthenticationManager authenticationManager;

	public UseServiceImpl(UserRepository repository, CollectRepository collectRepository,
		CollectCategoryRepository categoryRepository, NoticeRepository noticeRepository,
		EmailService emailService, PasswordEncoder passwordEncoder, AuthenticationManager authenticationManager) {
		this.repository = repository;
		this.categoryRepository = categoryRepository;
		this.collectRepository = collectRepository;
		this.noticeRepository = noticeRepository;
		this.emailService = emailService;
		this.passwordEncoder = passwordEncoder;
		this.authenticationManager = authenticationManager;
	}


	@Override
	public User loginByUsernameAndPassword(UsernamePasswordLoginView view) {
		var auth = new UsernamePasswordAuthenticationToken(view.getUsername(), view.getPassword());
		var validAuthToken = authenticationManager.authenticate(auth);
		SecurityContextHolder.getContext().setAuthentication(validAuthToken);

		var userDetails = (JwtUserDetails) validAuthToken.getPrincipal();
		return userDetails.getDelegateUser();
	}

	@Transactional
	@Override
	public User registerByEmail(EmailRegisterView view) {
		var user = new User();
		user.setNickname(view.getNickname());
		user.setUsername(view.getUsername());
		user.setEmail(view.getEmail());
		user.setPassword(passwordEncoder.encode(view.getPassword()));
		var result = repository.save(user);

		emailService.sendActivateEmail();

		return result;
	}

	@Transactional
	@Override
	public User activate(User user) {
		user.setActivateStatus(true);
		var result = repository.save(user);

		emailService.sendHelloEmail();

		return result;
	}

	@Transactional
	@Override
	public User resetPassword(User user, String newPassword) {
		user.setPassword(passwordEncoder.encode(newPassword));
		return repository.save(user);
	}

	@Transactional
	@Override
	public User update(Long id, User user) {
		var rawUser = repository.getOne(id);
		rawUser.setNickname(user.getNickname());
		rawUser.setIntroduce(user.getIntroduce());
		rawUser.setAvatarUrl(user.getAvatarUrl());
		rawUser.setBackgroundUrl(user.getBackgroundUrl());
		return repository.save(user);
	}

	@Cacheable("user")
	@Override
	public User get(Long id) {
		return repository.findById(id).orElseThrow(NotFoundException::new);
	}

	@Override
	public User getByRandom() {
		var count = repository.count();
		var randomId = RandomExtension.INSTANCE.range(1, count);
		return repository.findById(randomId).orElseThrow(NotFoundException::new);
	}

	@Cacheable("user.followToUserPage")
	@Override
	public Page<User> getFollowToUserPage(Long id, Pageable pageable) {
		return repository.findByFollowByUser_Id(id, pageable);
	}

	@Cacheable("user.followToUserCount")
	@Override
	public Long getFollowToUserCount(Long id) {
		return repository.countByFollowByUser_Id(id);
	}

	@Cacheable("user.followByUserPage")
	@Override
	public Page<User> getFollowByUserPage(Long id, Pageable pageable) {
		return repository.findByFollowToUser_Id(id, pageable);
	}

	@Cacheable("user.followByUserCount")
	@Override
	public Long getFollowByUserCount(Long id) {
		return repository.countByFollowToUser_Id(id);
	}

	@Cacheable("user.collectPage")
	@Override
	public Page<Collect> getCollectPage(Long id, Pageable pageable) {
		return collectRepository.findByUser_IdAndDeleteStatus(id, false, pageable);
	}

	@Cacheable("user.collectCount")
	@Override
	public Long getCollectCount(Long id) {
		return collectRepository.countByUser_IdAndDeleteStatus(id, false);
	}

	@Cacheable("user.collectCategoryPage")
	@Override
	public Page<CollectCategory> getCollectCategoryPage(Long id, Pageable pageable) {
		return categoryRepository.findByUser_Id(id, pageable);
	}

	@Cacheable("user.collectCategoryCount")
	@Override
	public Long getCollectCategoryCount(Long id) {
		return categoryRepository.countByUser_Id(id);
	}

	@Cacheable("user.noticePage")
	@Override
	public Page<Notice> getNoticePage(Long id, Pageable pageable) {
		return noticeRepository.findByUser_Id(id, pageable);
	}

	@Cacheable("collect.noticeCount")
	@Override
	public Long getNoticeCount(Long id) {
		return noticeRepository.countByUser_Id(id);
	}

	@Cacheable("collectPage")
	@Override
	public Page<User> findAll(Pageable pageable) {
		return repository.findAll(pageable);
	}

	@Cacheable("collectPage.byNickname")
	@Override
	public Page<User> findByNickname(String nickname, Pageable pageable) {
		return repository.findByNicknameContains(nickname, pageable);
	}

	@Cacheable("collectPage.byRole")
	@Override
	public Page<User> findByRole(Role role, Pageable pageable) {
		return repository.findByRole(role, pageable);
	}

	@Override
	public boolean exists(User user) {
		var username = user.getUsername();
		var email = user.getEmail();
		return repository.existsByUsernameOrEmail(username, email);
	}
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/service/impl/UseServiceImpl.kt
package com.windea.demo.cloudcollect.core.service.impl

import com.windea.commons.kotlin.extension.*
import com.windea.demo.cloudcollect.core.domain.entity.*
import com.windea.demo.cloudcollect.core.domain.enums.*
import com.windea.demo.cloudcollect.core.domain.model.*
import com.windea.demo.cloudcollect.core.domain.view.*
import com.windea.demo.cloudcollect.core.exception.*
import com.windea.demo.cloudcollect.core.repository.*
import com.windea.demo.cloudcollect.core.service.*
import org.springframework.cache.annotation.*
import org.springframework.data.domain.*
import org.springframework.security.authentication.*
import org.springframework.security.core.context.*
import org.springframework.security.crypto.password.*
import org.springframework.stereotype.*
import javax.transaction.*

@Service
@CacheConfig(cacheNames = ["user"])
open class UseServiceImpl(
	private val repository: UserRepository,
	private val emailService: EmailService,
	private val passwordEncoder: PasswordEncoder,
	private val authenticationManager: AuthenticationManager
) : UserService {
	override fun loginByUsernameAndPassword(view: UsernamePasswordLoginView): User {
		val authentication = UsernamePasswordAuthenticationToken(view.username, view.password)
		val validAuthentication = authenticationManager.authenticate(authentication)
		SecurityContextHolder.getContext().authentication = validAuthentication
		
		return (validAuthentication.principal as JwtUserDetails).delegateUser
	}
	
	@Transactional
	@CacheEvict(allEntries = true)
	override fun registerByEmail(view: EmailRegisterView): User {
		val user = User()
		user.nickname = view.nickname
		user.username = view.username
		user.email = view.email
		user.password = passwordEncoder.encode(view.password)
		val result = repository.save(user)
		
		emailService.sendActivateEmail()
		return result
	}
	
	@Transactional
	@CacheEvict(allEntries = true)
	override fun activate(user: User): User {
		user.isActivated = true
		val result = repository.save(user)
		
		emailService.sendHelloEmail()
		return result
	}
	
	@Transactional
	@CacheEvict(allEntries = true)
	override fun resetPassword(user: User, newPassword: String): User {
		user.password = passwordEncoder.encode(newPassword)
		return repository.save(user)
	}
	
	@Transactional
	@CacheEvict(allEntries = true)
	override fun update(id: Long, user: User): User {
		val rawUser = findById(id)
		rawUser.nickname = user.nickname
		rawUser.introduce = user.introduce
		rawUser.avatarUrl = user.avatarUrl
		rawUser.backgroundUrl = user.backgroundUrl
		return repository.save(user)
	}
	
	@Cacheable(key = "methodName + args")
	override fun findById(id: Long): User {
		return repository.findById(id).orElseThrow { NotFoundException() }
	}
	
	@Cacheable(key = "methodName + args")
	override fun findByUsername(username: String): User {
		return repository.findByUsername(username).orElseThrow { NotFoundException() }
	}
	
	@Cacheable(key = "methodName + args")
	override fun findByEmail(email: String): User {
		return repository.findByEmail(email).orElseThrow { NotFoundException() }
	}
	
	override fun findByRandom(): User {
		val count = repository.count()
		val randomId = RandomExtension.range(1, count)
		return repository.findById(randomId).orElseThrow { NotFoundException() }
	}
	
	@Cacheable(key = "methodName + args")
	override fun findAll(pageable: Pageable): Page<User> {
		return repository.findAll(pageable)
	}
	
	@Cacheable(key = "methodName + args")
	override fun findAllByNicknameContains(nickname: String, pageable: Pageable): Page<User> {
		return repository.findAllByNicknameContains(nickname, pageable)
	}
	
	@Cacheable(key = "methodName + args")
	override fun findAllByRole(role: Role, pageable: Pageable): Page<User> {
		return repository.findAllByRole(role, pageable)
	}
	
	@Cacheable(key = "methodName + args")
	override fun findAllByFollowToUserId(followToUserId: Long, pageable: Pageable): Page<User> {
		return repository.findAllByFollowToUserId(followToUserId, pageable)
	}
	
	override fun countByFollowToUserId(followToUserId: Long): Long {
		return repository.countByFollowToUserId(followToUserId)
	}
	
	@Cacheable(key = "methodName + args")
	override fun findAllByFollowByUserId(followByUserId: Long, pageable: Pageable): Page<User> {
		return repository.findAllByFollowByUserId(followByUserId, pageable)
	}
	
	override fun countByFollowByUserId(followByUserId: Long): Long {
		return repository.countByFollowByUserId(followByUserId)
	}
	
	@Cacheable(key = "methodName + args")
	override fun findAllByPraiseToCollectId(praiseToCollectId: Long, pageable: Pageable): Page<User> {
		return repository.findAllByPraiseToCollectId(praiseToCollectId, pageable)
	}
	
	override fun countByPraiseToCollectId(praiseToCollectId: Long): Long {
		return repository.countByPraiseToCollectId(praiseToCollectId)
	}
	
	override fun exists(user: User): Boolean {
		val username = user.username
		val email = user.email
		return repository.existsByUsernameOrEmail(username, email)
	}
}
